### PR TITLE
🚨 Always return Cache objects in modelings (to align with generate)

### PR DESCRIPTION
**Pull request by manueldeprada**
This PR removes reliance on `Cache.from_legacy_cache(past_key_values)` for initializing `None` `past_key_values`, replacing it with explicit cache initialization. The previous approach also set `return_legacy_cache=True`, unintentionally returning legacy tuples and masking other issues.

This change is necessary to support the upcoming deprecation of `from_legacy_cache` in v4.58.

Note: This update revealed an issue in `pipelines`, where `loader_batch_item` expects legacy tuples when iterating over `ModelOutputs`. It failed when encountering `Cache` objects.

### CODE REVIEW

**Review by gante — DISMISSED** [review]
LGTM! thank you for working on it

File: src/transformers/models/biogpt/modeling_biogpt.py [hunk 1]
```diff
@@ -553,15 +553,17 @@ def forward(
...
             return_legacy_cache = True
             logger.warning_once(
                 "Passing a tuple of `past_key_values` is deprecated and will be removed in Transformers v4.58.0. "
-                "You should pass an instance of `EncoderDecoderCache` instead, e.g. "
-                "`past_key_values=EncoderDecoderCache.from_legacy_cache(past_key_values)`."
+                "You should pass an instance of `DynamicCache` instead, e.g. "
+                "`past_key_values=DynamicCache.from_legacy_cache(past_key_values)`."
             )
-            past_key_values = EncoderDecoderCache.from_legacy_cache(past_key_values)
+            past_key_values = DynamicCache.from_legacy_cache(past_key_values)
```
Comment (Cyrilvallez) [inline src/transformers/models/biogpt/modeling_biogpt.py:567]: Damn, indeed it was the wrong type of Cache here... 🥲🤯

File: src/transformers/cache_utils.py [hunk 2]
```diff
@@ -1730,15 +1732,15 @@ class EncoderDecoderCache(Cache):
     # Override @property from Cache
     is_compileable = None
 
-    def __init__(self, self_attention_cache: Cache, cross_attention_cache: Cache):
+    def __init__(self, self_attention_cache: Cache = None, cross_attention_cache: Cache = None):
```
Comment (Cyrilvallez) [inline src/transformers/cache_utils.py]: Let's not add a default here, and do it in the models themselves instead! Because the `__init__` will change in https://github.com/huggingface/transformers/pull/40008, and https://github.com/huggingface/transformers/pull/40039 provides support for passing the `config` in `DynamicCache` as well, which could be done in both instances of the 2 caches for EncoderrDecoderCache

**Review by Cyrilvallez — COMMENTED** [review]
Hey, sorry for the delay on this! Thanks a lot, happy to start cleaning up everything to finally drop the legacy format soon! This needs a rebase to fix the conflict though! And for EncoderDecoderCache, let's not provide default values to the __init__, let's instantiate with 2 DynamicCache in the models instead (see the 2 PR I linked!)

File: src/transformers/cache_utils.py (hunk 2 above)
Comment (manueldeprada) [reply src/transformers/cache_utils.py]: awesome, done!

**Review by Cyrilvallez — DISMISSED** [review]
Perfect, thanks a lot! Looks like you just need to run `make fix-copies` for Flaubert (probably only the docstrings change) to make CI happy.
If not consistent anymore based on real change, then you can just break the copy!
Feel free to merge once it's done and CI is happy!

**Comment by gante** [comment]
@manueldeprada @Cyrilvallez sorry to potentially add one more task here :D Should we update all `DynamicCache()` added in this PR to `DynamicCache(config=self.config)`, following the pattern introduced in #40039 ?

**Comment by manueldeprada** [comment]
There was a few more models that were using `from_legacy_cache` as init (and, in consequence, returning a legacy tuple in `forward()`)

Fixing them exposed a larger problem: **`Trainer` cannot handle non-tuple model outputs**.

Example:
`tests/trainer/test_trainer.py::TrainerIntegrationTest::test_evaluation_with_keys_to_drop` expects `logits` from the prediction loop to be a tuple with logits and also `past_key_values`. This requires `Trainer` to handle Caches, including this line:
https://github.com/huggingface/transformers/blob/41980ce93e775f6c88500c51c8db7946fc6a2add/src/transformers/trainer.py#L4549

That calls `accelerate/utils/operations.py:recursively_apply` or `src/transformers/trainer_pt_utils.py:nested_concat` which were able to process tuples of tensors, but not Caches.

I tried forcing `Trainer` to use the legacy cache format ([commit](https://github.com/huggingface/transformers/pull/39765/commits/0ad4c8f8675e003b0677220c48961180d162f714)), but that broke more tests and its bad code: some models dont even use `past_key_values`.

I need some guidance on how to properly handle this, as I am also not familiar with Trainer. Should we just disallow using logits to return `past_key_values`?  cc @gante @Cyrilvallez

**Comment by Cyrilvallez** [comment]
@gante indeed it would be nice to start using the config in all cache initialization, I was a bit too lazy before (just made sure that it was there for all sliding window/hybrid models) but it's indeed planned. Probably we can do it in a follow-up PR with simple regex match on all modeling files though, in order to simplify this PR.
@manueldeprada I think we purely want to drop Trainer using past key values here, not fix it - especially as it does not seem to be used (it's not supported on any recent models which have been returning a Cache for a longggg time) -> you can simply remove the test

**Comment by Cyrilvallez** [comment]
Generally, we don't want any link Trainer <-> Cache as caching is only an inference speedup trick

File: src/transformers/cache_utils.py [hunk 3]
```diff
@@ -131,6 +132,8 @@ def get_seq_length(self) -> int:
         """Returns the sequence length of the cached states."""
         if self.keys is None or self.keys.numel() == 0:
             return 0
+        if len(self.keys.shape) < 2:
+            return 0
         return self.keys.shape[-2]
```
Comment (Cyrilvallez) [inline src/transformers/cache_utils.py]: wait what is this?

File: src/transformers/cache_utils.py [hunk 4]
```diff
@@ -159,9 +162,15 @@ def batch_repeat_interleave(self, repeats: int) -> None:
...
+        self.select_indices(0, indices)
+
+    def select_indices(self, axis: int, indices: torch.Tensor) -> None:
+        """Only keep the `indices` in the selected dimension of the cache."""
         if self.get_seq_length() > 0:
-            self.keys = self.keys[indices, ...]
-            self.values = self.values[indices, ...]
+            indices = torch.as_tensor(indices, dtype=torch.long, device=self.keys.device)
+            self.keys = self.keys.index_select(dim=axis, index=indices)
+            self.values = self.values.index_select(dim=axis, index=indices)
```
Comment (Cyrilvallez) [inline src/transformers/cache_utils.py]: And here why do we do this?

File: src/transformers/cache_utils.py [hunk 5]
```diff
@@ -905,9 +914,15 @@ def batch_repeat_interleave(self, repeats: int):
...
+        self.select_indices(0, indices, return_copy=False)
+
+    def select_indices(self, axis: int, indices: torch.Tensor, return_copy: bool = False):
         """Select indices from the cache"""
-        for layer_idx in range(len(self.layers)):
-            self.layers[layer_idx].batch_select_indices(indices)
+        cache = copy.deepcopy(self) if return_copy else self
+        for layer_idx in range(len(cache.layers)):
+            cache.layers[layer_idx].select_indices(axis, indices)
+        return cache
```
Comment (Cyrilvallez) [inline src/transformers/cache_utils.py]: It's an extremely bad idea to copy the cache!!!

**Review by Cyrilvallez — COMMENTED** [review]
Wait, why do we have unrelated changes to the cache here?

File: tests/utils/test_cache_utils.py [hunk 6]
```diff
@@ -1197,6 +1197,28 @@ def test_dynamic_cache(self):
             "DynamicCache Scenario 2 layer 1 failed",
         )
 
+    def test_dynamic_cache_batch_select_indices(self):
```
Comment (manueldeprada) [inline tests/utils/test_cache_utils.py:1200]: This additional test does not hurt, since `batch_select_indices` was never tested

File: src/transformers/pipelines/pt_utils.py [hunk 7]
```diff
@@ -86,12 +86,16 @@ def loader_batch_item(self):
...
+                if k in {"hidden_states", "past_key_values", "attentions"}:
+                    if isinstance(element, tuple):
+                        # Those are stored as lists of tensors so need specific unbatching.
+                        if isinstance(element[0], torch.Tensor):
+                            loader_batched[k] = tuple(el[self._loader_batch_index].unsqueeze(0) for el in element)
+                        elif isinstance(element[0], np.ndarray):
+                            loader_batched[k] = tuple(
+                                np.expand_dims(el[self._loader_batch_index], 0) for el in element
+                            )
+                    # Cache objects are skipped here
```
Comment (manueldeprada) [inline src/transformers/pipelines/pt_utils.py]: check this, unsure on this part on the library

File: src/transformers/cache_utils.py [hunk 8]
```diff
@@ -1063,6 +1063,8 @@ def from_legacy_cache(cls, past_key_values: tuple[tuple[torch.Tensor, torch.Tens
         backward compatibility.
         """
         cache = cls()
+        if past_key_values is None:
+            logger.warning_once("past_key_values should be None in from_legacy_cache()")
```
Comment (manueldeprada) [inline src/transformers/cache_utils.py]: CI was green with a `raise ValueError` here, but for release a warning is better so we don't break code on the Hub

Comment (Cyrilvallez) [reply src/transformers/cache_utils.py]: Alright, but missing a negation here

File: examples/modular-transformers/modeling_dummy_bert.py [hunk 9]
```diff
@@ -540,8 +540,13 @@ def forward(
                 )
                 use_cache = False
 
+        if use_cache and self.config.is_decoder and past_key_values is None:
+            past_key_values = EncoderDecoderCache(
+                self_attention_cache=DynamicCache(), cross_attention_cache=DynamicCache()
+            )
```
Comment (Cyrilvallez) [inline examples/modular-transformers/modeling_dummy_bert.py]: Let's not use `kwargs` in EncoderDecoderCache's __init__, see https://github.com/huggingface/transformers/pull/40008
```suggestion
            past_key_values = EncoderDecoderCache(DynamicCache(), DynamicCache())
```

File: src/transformers/pipelines/pt_utils.py (hunk 7 above)
Comment (Cyrilvallez) [inline src/transformers/pipelines/pt_utils.py]: Let's not change previous logic, but simply make it very explicit that they are skipped

```suggestion
                if k in {"hidden_states", "attentions"} and isinstance(element, tuple):
                    # Those are stored as lists of tensors so need specific unbatching.
                    if isinstance(element[0], torch.Tensor):
                        loader_batched[k] = tuple(el[self._loader_batch_index].unsqueeze(0) for el in element)
                    elif isinstance(element[0], np.ndarray):
                        loader_batched[k] = tuple(np.expand_dims(el[self._loader_batch_index], 0) for el in element)
                    continue
                if k == "past_key_values":
                    continue
```

File: tests/generation/test_utils.py [hunk 10]
```diff
@@ -2043,7 +2043,12 @@ def test_generate_with_quant_cache(self):
                 model.generate(past_key_valyes=DynamicCache(), **generation_kwargs, **inputs_dict)
 
             # setting incorrect cache_config args should raise an Error, i.e. nbits=60 does not make sense
-            generation_kwargs["cache_config"] = {"nbits": 60, "q_group_size": 8, "residual_length": 128}
+            generation_kwargs["cache_config"] = {
+                "backend": "quanto",
+                "nbits": 60,
+                "q_group_size": 8,
+                "residual_length": 128,
+            }
```
Comment (Cyrilvallez) [inline tests/generation/test_utils.py]: This should should not be needed and is unrelated, let's revert!

File: tests/trainer/test_trainer.py [hunk 11]
```diff
@@ -1326,10 +1326,8 @@ def test_evaluation_with_keys_to_drop(self):
         # By default the past_key_values are removed
         result = trainer.predict(eval_dataset)
         self.assertTrue(isinstance(result.predictions, np.ndarray))
-        # We can still get them by setting ignore_keys to []
-        result = trainer.predict(eval_dataset, ignore_keys=[])
-        self.assertTrue(isinstance(result.predictions, tuple))
-        self.assertEqual(len(result.predictions), 2)
+        # Note: past_key_values must be ignored, as Trainer cannot handle Cache classes
+        # and support for legacy tuples is deprecated.
```
Comment (Cyrilvallez) [inline tests/trainer/test_trainer.py]: Let's remove the whole test here, otherwise just keeping everything except that does not make sense (see the name of the test)

**Review by Cyrilvallez — COMMENTED** [review]
Alright, thanks for reverting previous changes. 
We just need to remove the `kwarg` in all the initialization of `EncoderDecoderCache` to simplify our lives with https://github.com/huggingface/transformers/pull/40008.
Also, I just checked `generate`, and we actually return `Cache` format all the time, even when `kv` are provided as `tuple`. So, I think it makes sense to do the same here, and always return a Cache, whatever the input, not only half the time. So we can drop `return_legacy_cache` in all modeling, and never go back to legacy format 🤗

**Comment by manueldeprada** [comment]
@Cyrilvallez all done!! thanks, now the diff is very nice: +378 −538 🧹 🧹 😄

**Comment by manueldeprada** [comment]
run-slow: autoformer, bark, bart, bert, bert_generation, big_bird, bigbird_pegasus, biogpt, blenderbot, blenderbot_small, blip, bridgetower, camembert, clvp, cpmant, ctrl

**Comment by manueldeprada** [comment]
Checked the slow tests and all the failures seem unrelated. There are a bunch of them that will be fixed by #40008 !

**Review by Cyrilvallez — APPROVED** [review]
LGTM! Thanks a lot for the cleanup! Happy to get rid of it! Feel free to merge! 🤗

**Comment by gante** [comment]
> Generally, we don't want any link Trainer <-> Cache as caching is only an inference speedup trick

@Cyrilvallez for completeness, PEFT has some fine-tuning methods using caches -- which is why we don't simply set `use_cache &= not self.training`. AFAIK Trainer has no use for caches, except in the eval step.

**Review by gante — APPROVED** [review]
LGTM, thank you for iterating 💛

**Comment by Cyrilvallez** [comment]
Nice, thanks @gante for the heads-up! cc @BenjaminBossan then just in case, but I believe peft already supports both formats anyway, as recent models have been returning Cache objects for a long time! 🤗

**Comment by BenjaminBossan** [comment]
Thanks for the heads up, I ran the relevant tests against the latest transformers main branch and they all passed. The new warning was not triggered, so I think we're good on the PEFT side.
